### PR TITLE
HARP-6512 Remove linear tile search from TerrainElevationProvider

### PR DESCRIPTION
**Pull request by atomicsulfate**
getDisplacementMap and getHeight methods in TerrainElevationProvider do a linear search over all terrain tiles cached in the VisibleTileSet until the tile with the requested key/geopoint is found.

This change improves this by getting directly the proper tile from the LRU tile cache.
For getHeight, the tile key is calculated from the given geopoint using terrainSource.getTilingScheme().getTileKey().

### CODE REVIEW

File: @here/harp-mapview/lib/VisibleTileSet.ts [hunk 1]
```diff
@@ -638,6 +631,51 @@ export class VisibleTileSet {
...
+        function updateTile(tileToUpdate?: Tile) {
+            if (tileToUpdate === undefined) {
+                return;
+            }
+            // Keep the tile from being removed from the cache.
+            tileToUpdate.frameNumLastRequested = dataSource.mapView.frameNumber;
+        }
+
+        if (!dataSource.cacheable) {
+            const resultTile = dataSource.getTile(tileKey);
```
Comment (ninok) [inline @here/harp-mapview/lib/VisibleTileSet.ts]: I think we should not call `dataSource.getTile(tileKey)` if `cacheOnly===true`

Comment (atomicsulfate) [reply @here/harp-mapview/lib/VisibleTileSet.ts]: done

**Comment by atomicsulfate** [comment]
@heremaps-bot retest this please

File: @here/harp-mapview/lib/VisibleTileSet.ts [hunk 2]
```diff
@@ -364,40 +372,133 @@ export class VisibleTileSet {
...
+    }
+
+    /**
+     * Gets the tile corresponding to the given data source and location from the rendered tiles.
+     *
+     * @param dataSource The data source the tile belongs to.
+     * @param geoPoint The geolocation included within the tile.
+     * @return The tile if found among the rendered tiles, undefined otherwise.
+     */
+    getRenderedTileAtLocation(
```
Comment (nzjony) [inline @here/harp-mapview/lib/VisibleTileSet.ts]: Should have a test added, but can be in a separate PR.

File: @here/harp-mapview/lib/VisibleTileSet.ts [hunk 3]
```diff
@@ -364,40 +372,133 @@ export class VisibleTileSet {
...
+     * @param tileKey The key identifying the tile.
+     * @param offset Tile offset.
+     * @return The tile if found among the rendered tiles, undefined otherwise.
+     */
+    getRenderedTile(
+        dataSource: DataSource,
+        tileKey: TileKey,
+        offset: number = 0
+    ): Tile | undefined {
+        const dataSourceVisibleTileList = this.dataSourceTileList.find(list => {
```
Comment (nzjony) [inline @here/harp-mapview/lib/VisibleTileSet.ts]: Maybe this could be a map to simplify things a bit.

File: @here/harp-mapview/lib/text/TextElementsRenderer.ts [hunk 4]
```diff
@@ -2201,21 +2201,21 @@ export class TextElementsRenderer {
     }
 
     private renderTileList(
-        visibleTiles: Tile[],
+        visibleTiles: Map<number, Tile>,
```
[the remark this answers is not in the document]
Comment (atomicsulfate) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: visibleTiles.values() is an iterator, and you can't get the size from it, which is needed in line 2212.

[the remark this answers is not in the document]
Comment (nzjony) [reply @here/harp-mapview/lib/text/TextElementsRenderer.ts]: Ok

**Comment by atomicsulfate** [comment]
@heremaps-bot retest this please
